### ISSUE TITLE
Not possible to run custom inventory plugin as inventory

### ISSUE THREAD

**Issue by machacekondra** [ansible/awx#3113]
##### ISSUE TYPE
 - Feature Idea

##### COMPONENT NAME

 - API
 - UI

##### SUMMARY

I couldn't setup a project with using my custom inventory plugin. Can you please describe or document, how could I use custom inventory plugin. Using dynamic inventory script is simple and straightforward, but not inventory plugins.

**Comment by AlanCoding** [comment]
Yes, you need this patch to Ansible:

https://github.com/ansible/ansible/pull/51177

The recommended way is ("will be" might be more accurate) that you use the special named folder adjacent to your inventory file inside of source control. Then you have this imported into AWX via an inventory source sourced from a project. Ansible has several special names where it knows to look for such, such as `group_vars` or `host_vars` folders for example. In this case, Ansible knows to search the adjacent `inventory_plugins` folder for custom inventory plugins.

If at all possible, you should also write your plugin such that it can work correctly via the auto plugin. That means that you would have an inventory file like `steve.yml`, and then it would have contents looking like:

```yaml
plugin: steve
param1: value
param2: another_value
```

Then the logic within your inventory plugin will receive those parameters and know what to do with them.

If you hate the auto plugin, then you have to set "steve" as an enabled plugin. To do this, you'll probably have to use an adjacent `ansible.cfg` file to add "steve" to the list of enabled inventory plugins. The auto plugin avoids this extra work. There might be some more nuances with expected paths for the config file, which have not yet been anticipated, and this approach might be unstable for a while to come.

**Comment by AlanCoding** [comment]
Forgot to mention, once that patch is done in Ansible, we do need to add --playbook-dir option on our side.

**Comment by AlanCoding** [comment]
This is implemented on our side now. It's waiting on Ansible upstream now.

Closing, since there is nothing else to do, tracked elsewhere.

**Comment by fculpo** [comment]
Would it be possible to pass extra vars to the inventory plugin ?

Like:
```
plugin: steve
param1: "{{ var1 }}"
param2: "{{ var2 }}"
```
This is doable in Ansible core

**Comment by AlanCoding** [comment]
Could you give more details on that use case? Do you mean that you have an inventory file in source control with those var1 & var2 templated entries? What Ansible command, specifically, works to accomplish that on the CLI for you?

**Comment by fculpo** [comment]
I could not do it in CLI eitheir, i was really thinking this would work.
I want to pass parameters to the .yml inventory config file at inventory launch time, so that my plugin can use them.

I tried something like:

```ansible-playbook -i steve.yml someplaybook.yml -e "var1=anything var2=anything"```

with steve.yml =
```
plugin: steve
param1: "{{ var1 }}"
param2: "{{ var2 }}"
```

and an inventory plugin named steve.py

But if I output param1 and param2 in the inventory plugin, i get:
param1 = {{ var1 }}
param2 = {{ var2 }}

**Comment by AlanCoding** [comment]
It's possible that templating logic would need to be added to the inventory plugin code, `steve.py`, because it is not part of the base plugin class. A more interesting question is whether there is any clean way for that code to obtain data from extra_vars. If not, then AWX can't consume something that Ansible core doesn't support. If it is possible, then we could consider adding the logic to the base plugin class, and then nothing would be stopping AWX from supporting use of extra_vars in inventory updates in various forms.

**Comment by megakoresh** [comment]
@AlanCoding With AWX 7.0.0 when selecting inventory source option as sourced from project, the file selector does not detect yaml file that references a custom inventory plugin (panorama). The directory structure is like this

- root
  - panos
    - inventories
      - panorama
        - inventory.yml <- this is the yml file with `plugin: panorama`
    - search-fw-rules.yml <- playbook
    - inventory_plugins
      - panorama.py <- inventory plugin

Running `ansible-inventory` with `--playbook-dir` option works just fine locally on ansible `2.8.4`:

```sh
ansible-inventory -i panos/inventories/panorama --playbook-dir=$(pwd)/panos --list
```

But in `awx` the file select list does not allow me to select `panorama` or `inventory.yml` as an inventory source. It only shows executable scripts or static inventory files (e.g. `panos/inventories/statis/hosts`)

Edit: I can manually type in the path to that field, which makes it parse the inventory, however it then doesn't detect the plugin because, presumably, it's not at the root of the repository:

```
panos/inventories/panorama/inventory.yml' specifies unknown plugin 'panorama'
```

How does awx determine what path to pass with `--playbook-dir` option?

**Comment by AlanCoding** [comment]
`--playbook-dir` is the directory the inventory file is in.

https://github.com/ansible/awx/blob/806648af8963c9ac3fb6f13c239a62e89894aab4/awx/main/management/commands/inventory_import.py#L66-L68

https://github.com/ansible/awx/blob/806648af8963c9ac3fb6f13c239a62e89894aab4/awx/main/management/commands/inventory_import.py#L81

https://github.com/ansible/awx/blob/806648af8963c9ac3fb6f13c239a62e89894aab4/awx/main/management/commands/inventory_import.py#L142

In your case, that is `root/panos/inventories/panorama`. I get that your file structure would work if the playbook dir was set to `root/panos`, but this is neither the top-level project directory (the working directory in job runs) or the inventory's directory, so I don't see how AWX would be able to get this information.

It is _thinkable_ that you can use `ansible.cfg` in your inventory directory (which again, is set as the playbook dir) to point to a directory that contains the magic directory. This is too complicated to explain in a comment. Here:

https://github.com/AlanCoding/Ansible-inventory-file-examples/commit/b3f559d0c4d07f1e176ac5a148bbf87164fde859

I also don't see any hazard in doing this under the conditions that (1) your inventory directory only has that single inventory and (2) you don't intend to source your plugin from different places. Both of these conditions are probably met.

**Comment by AlanCoding** [comment]
> But in awx the file select list does not allow me to select panorama or inventory.yml as an inventory source.

That is https://github.com/ansible/awx/pull/4171

**Comment by megakoresh** [comment]
> In your case, that is `root/panos/inventories/panorama`. I get that your file structure would work if the playbook dir was set to `root/panos`, but this is neither the top-level project directory (the working directory in job runs) or the inventory's directory, so I don't see how AWX would be able to get this information.

Well in my opinion the correct approach would be to let me specify this parameter myself.

> It is _thinkable_ that you can use `ansible.cfg` in your inventory directory (which again, is set as the playbook dir) to point to a directory that contains the magic directory.

I don't think this would work because there are multiple different folders in that repo each with their own inventories and playbooks and `ansible.cfg`. You would probably suggest that these be split into separate repositories, which would solve the issue, but all of those playbooks are used together by same people, they are split by vendor since working with network equipment all of those are always very different. The full structure is actually

- root (by root I mean repo root btw)
  - panos
    - inventories
      - panorama
        - inventory.yml
    - inventory_plugins
    - some-playbooks.yml
  - nxos
    - inventories
      - nexus9
        - hosts
      - mgmt
    - some-playbooks.yml
- ...other vendors with same structure

For now I will split it to separate repository as a workaround, but adding an option to set `playbook-dir` manually would be a better option IMO. 

In fact setting the the playbook directory to the same directory as the inventory is against ansible conventions in the first place, since a _directory_ is a valid inventory source in ansible and a much more common use-case than simply dropping `hosts` file alongside playbooks.

**Comment by AlanCoding** [comment]
I don't see why your take away is splitting your stuff into separate repos.

> I don't think this would work because there are multiple different folders in that repo each with their own inventories and playbooks and ansible.cfg.

Maybe you were looking at the wrong case from my examples. The case I intended to correspond to your scenario was

https://github.com/AlanCoding/Ansible-inventory-file-examples/tree/master/plugins/config_path/inventory_dir

see `ansible.cfg`

```
[defaults]
inventory_plugins=../other_dir/inventory_plugins
```

This config would not get used by your playbook run itself, and it should apply fairly specifically to your AWX inventory updates.

**Comment by megakoresh** [comment]
> Maybe you were looking at the wrong case from my examples. The case I intended to correspond to your scenario was
>
> https://github.com/AlanCoding/Ansible-inventory-file-examples/tree/master/plugins/config_path/inventory_dir

Oh that's how it works. I was looking at wrong example. Ok, I didn't even know you can apply ansible configuration this way. Allowing users to specify `playbook-dir` for sources still seems like a cleaner and more robust solution than some magical detection, but your example does solve this particular use-case.

So thanks!

**Comment by AlanCoding** [comment]
Sorry, I think I was wrong in the suggestion. The playbook dir is set to the inventory directory, but `ansible.cfg` comes from the working directory and not the playbook directory. IIRC, working directory is the root of the project for inventory updates and jobs. So you could point to the directory with the inventory_plugins directory from that config file, but if that conflicts with some other use then that would be very difficult to get around.

**Comment by megakoresh** [comment]
> Sorry, I think I was wrong in the suggestion. The playbook dir is set to the inventory directory, but `ansible.cfg` comes from the working directory and not the playbook directory. IIRC, working directory is the root of the project for inventory updates and jobs. So you could point to the directory with the inventory_plugins directory from that config file, but if that conflicts with some other use then that would be very difficult to get around.

Yes, it would conflict with basically any other use-case. And yes I haven't been able to get your suggestion to work. So my initial assumptions were correct. 

It's very strange to me that AWX has this custom logic for problems which have for a long time been solved in the cli interface. I think ansible semaphore has a better architecture in that respect.

The purpose of AWX is to provide governance and web GUI to ansible, and coding some kind of custom logic to bypass standard ansible behaviour sounds for things like directory resolution is like working against yourself. I will open at least a feature request for adding playbook option to sources form. Not sure if I have time at the moment to do a PR though. Not at the moment at least.

**Comment by Throckmortra** [comment]
Hey @AlanCoding I'm unable to get your example to work here. It seems that AWX just can't load my custom plugins regardless of where I place them relative to the working directory or specifying them with that ansible.cfg hack. Is there any documentation available for getting dynamic inventory plugins working in AWX? It still claims they're "unknown"

**Comment by AlanCoding** [comment]
Do you mean https://github.com/AlanCoding/Ansible-inventory-file-examples/tree/master/plugins/config_path/inventory_dir specifically? I can re-test that if it has actually regressed.

**Comment by Throckmortra** [comment]
@AlanCoding yes

**Comment by Throckmortra** [comment]
my directory looks like this
```
inventory/staging/
├── hosts
├── my_plugin.yml 
├── ansible.cfg
inventory_plugins/
├── my_plugin.py
```


contents of ansible.cfg:
```
[defaults]
inventory_plugins=../../inventory_plugins
```

**Comment by Throckmortra** [comment]
@AlanCoding interestingly, this ansible.cfg hack works locally - I don't have to use `--playbook-dir=./` when running ansible-inventory from root. AWX doesn't seem to recognize it though

**Comment by Throckmortra** [comment]
Hey other people who find yourself here,

My (unfortunate) workaround for this issue was to create an installer playbook for AWX that I can run to "install" my inventory plugins

```
- name: copy all python plugins
  copy:
    src: "{{ item }}"
    dest: "/usr/share/ansible/plugins/inventory/"
    owner: "root"
    mode: 0755
   with_fileglob:
    - "inventory_plugins/*.py"
```

You can continue to keep your inventory plugins in the correct "magic" directory in the root of your project for Ansible CLI use. I think after researching this, AWX just doesn't have decent support for magic directories and none of the supposed hacks really work or feel better than this option.

**Comment by AlanCoding** [comment]
Okay, I see. This is because the working directory is set to the project folder root. As such, anything run out of the project will use the `ansible.cfg` for the project generally.

https://github.com/ansible/awx/blob/ddb8c93f1790a9f72db25a241bd57b2eb645fafb/awx/main/tasks.py#L2646-L2647

Yeah, that's painful for modifying the configuration, but jobs work this way, so people wouldn't expect that inventory would work differently. This could be changed to the directory of the inventory file, but I just don't know what else that would mess up.

So the current state of affairs is that you can specify the plugin in the project's top-level `ansible.cfg`, but not in a more targeted way along-side the inventory file.

I don't expect this solves your problem, but I am taking stock of what is currently happening.
